File: BetKeeper/src/DB/Tables/Bets.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Globalization;
using BetKeeper.Exceptions;

namespace BetKeeper.DB.Tables
{
    /// <summary>
    /// handles database operations for table bets.
    /// </summary>
    public class Bets
    {
        /// <summary>
        /// Returns all rows from table bets which match the query.
        /// </summary>
        /// <param name="user">If inputted, only bets from user with given username are searched.</param>
        /// <param name="bet_won"> If not null, used to return bets which have finished or not.</param>
        /// <param name="ConnectionString">Connectionstring to be used.</param>
        public static List<Bet> GetBets(string ConnectionString, int user = -1, bool? bet_finished = null)
        {
            string query = "SELECT * FROM bets ";

            if (user != -1)
                query = query + "WHERE owner = @owner ";

            if (user != -1 && bet_finished != null)
                if (bet_finished == true)
                    query = query + String.Format("AND bet_won != -1");
                else if (bet_finished == false)
                    query = query + String.Format("AND bet_won == -1");
            else if (bet_finished != null)
            {
                if (bet_finished == true)
                    query = query + String.Format("WHERE bet_won != -1");
                else if (bet_finished == false)
                    query = query + String.Format("AND bet_won == -1");
            }

            query = query + ";";
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();

            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("owner", user);
            SQLiteDataReader reader = command.ExecuteReader();

            List<Bet> result = new List<Bet>();

            while (reader.Read())
            {
                Bet obj = new Bet(Convert.ToInt32(reader["owner"].ToString()), reader["name"].ToString(), reader["date_time"].ToString(), Convert.ToInt32(reader["bet_won"].ToString()),
                    Convert.ToInt32(reader["bet_id"].ToString()), Convert.ToDouble(reader["odd"].ToString()),
                    Convert.ToDouble(reader["bet"].ToString()));

                result.Add(obj);
            }
            con.Close();

            return result;
        }

        /// <summary>
        /// Returns a bet object.
        /// Returns null if no bets matched or query returned more than one bet.
        /// </summary>
        public static Bet GetBet(string ConnectionString, int id)
        {
            string query = "SELECT * FROM bets WHERE bet_id = @bet_id;";
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();

            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("bet_id", id);
            SQLiteDataReader reader = command.ExecuteReader();

            List<Bet> bets = new List<Bet>();

            while (reader.Read())
            {
                Bet obj = new Bet(Convert.ToInt32(reader["owner"].ToString()), reader["name"].ToString(), reader["date_time"].ToString(), Convert.ToInt32(reader["bet_won"].ToString()),
                    Convert.ToInt32(reader["bet_id"].ToString()), Convert.ToDouble(reader["odd"].ToString()),
                    Convert.ToDouble(reader["bet"].ToString()));

                bets.Add(obj);
            }
            con.Close();

            return (bets.Count == 1) ? bets[0] : null;
        }

        /// <summary>
        /// returns all bets in specific folder.
        /// </summary>
        /// <param name="folder">Name of the folder.</param>
        /// <param name="owner">Folder owner id.</param>
        /// <param name="finished">Optional parameter, if not null, statement returns bets from folder
        /// that are either finished or not.</param>
        /// <param name="connectionString">Connectionstring used to connect to database.</param>
        /// <returns>Bets from the folder that match the inputted parameters.</returns>
        public static List<Bet> GetBetsInBetFolder(string ConnectionString, string folder, int owner, bool? bet_finished = null)
        {
            string query = "";
            if (bet_finished == null)
                query = "SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = @owner and bf.folder = @folder;";
            else if (bet_finished == true)           
                query = "SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = @owner and bf.folder = @folder and bet_won != -1;";
            else if (bet_finished == false)
                query = "SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = @owner and bf.folder = @folder and bet_won = -1;";

            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();

            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("owner", owner);
            command.Parameters.AddWithValue("folder", folder);
            SQLiteDataReader reader = command.ExecuteReader();

            List<Bet> result = new List<Bet>();

            while (reader.Read())
            {
                Bet obj = new Bet(Convert.ToInt32(reader["owner"].ToString()), reader["name"].ToString(), reader["date_time"].ToString(), Convert.ToInt32(reader["bet_won"].ToString()),
                    Convert.ToInt32(reader["bet_id"].ToString()), Convert.ToDouble(reader["odd"].ToString()),
                    Convert.ToDouble(reader["bet"].ToString()));

                result.Add(obj);
            }
            con.Close();

            return result;
        }

        /// <summary>
        /// Adds a new bet to the bets table.
        /// </summary>
        /// <returns>Id of the last inserted row.</returns>
        /// <exception cref="ArgumentException">Thrown when datetime parameter could not be converted into a datetime object.</exception>
        /// <exception cref="SQLiteException">Thrown if foreign key constraints fail</exception>
        public static int CreateBet(string ConnectionString, int user_id, string datetime, double odd, double bet, string name, bool? bet_won)
        {
            int result = -1;
            CultureInfo customCulture = (CultureInfo)System.Threading.Thread.CurrentThread.CurrentCulture.Clone();
            customCulture.NumberFormat.NumberDecimalSeparator = ".";
            System.Threading.Thread.CurrentThread.CurrentCulture = customCulture;
            DateTime date;

            if (bet_won != null)
                result = Convert.ToInt32(bet_won);

            if (DateTime.TryParse(datetime, CultureInfo.CurrentCulture, DateTimeStyles.AssumeUniversal, out date))
                datetime = date.ToString("yyyy-MM-dd H:mm:ss");
            else
                throw new ArgumentException("Date inputted is not valid");

            string query = "INSERT INTO bets (bet_won, name, odd, bet, date_time, owner) values (@bet_won, @name, @odd, @bet, @date_time, @owner);";

            int queryResult;
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();
            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("bet_won", result);
            command.Parameters.AddWithValue("name", name);
            command.Parameters.AddWithValue("odd", odd);
            command.Parameters.AddWithValue("bet", bet);
            command.Parameters.AddWithValue("date_time", datetime);
            command.Parameters.AddWithValue("owner", user_id);
            try
            {
                command.ExecuteNonQuery();
                queryResult = (int)con.LastInsertRowId;   
            }
            finally
            {
                con.Close();
            }
            return queryResult;
        }

        /// <summary>
        /// Adds bet to specific folders. Ignores folders which do not belong to selected user.
        /// </summary>
        /// <returns>Number of inserted rows.</returns>
        /// <exception cref="UnknownBetError">Thrown when bet with bet_id 'id' is not found.</exception>
        /// <exception cref="SQLiteException">Thrown if foreign key constraints fail</exception>
        public static int AddBetToFolders(string ConnectionString, List<string> folders, int bet_id, int user_id)
        {
            if (GetBet(ConnectionString, bet_id) == null)
                throw new UnknownBetError("Bet with id " + bet_id + " was not found");

            string query = "";
            List<string> user_folders = Folders.GetUsersFolders(ConnectionString, user_id);
            foreach(string folder in folders)
            {
                if (user_folders.Contains(folder))
                    query = query + String.Format("INSERT INTO bet_in_bet_folder VALUES ('{0}', @owner, @bet_id);", folder);
            }

            int queryResult;
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();
            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("owner", user_id);
            command.Parameters.AddWithValue("bet_id", bet_id);
            try
            {
                queryResult = command.ExecuteNonQuery();
            }
            finally
            {
                con.Close();
            }
            return queryResult;
        }

        /// <summary>
        /// Modifies a bet which matches bet_id. Can modify result, odd, bet, and name for the bet.
        /// Returns number of affected rows. 
        /// </summary>
        /// <exception cref="UnknownBetError">Thrown when bet with bet_id 'id' is not found.</exception>
        /// <exception cref="AuthenticationError">Thrown when user tries to modify another user's bet.</exception>
        public static int ModifyBet(string ConnectionString, int bet_id, int user_id, bool? bet_won, double? odd = null, double? bet = null, string name = null)
        {
            int result = -1;
            if (GetBet(ConnectionString, bet_id) == null)
                throw new UnknownBetError("Bet with id " + bet_id + " was not found");

            if (GetBet(ConnectionString, bet_id).getOwner() != user_id)
                throw new AuthenticationError("Bet did not belong to user trying to modify it");

            if (bet_won != null)
                result = Convert.ToInt32(bet_won);

            string query = "UPDATE bets SET bet_won = @bet_won";

            if (odd != null)
                query = query + ", odd = @odd";
            if (bet != null)
                query = query + ", bet = @bet";
            if (!String.IsNullOrEmpty(name))
                query = query + ", name = @name";

            query = query + String.Format(" WHERE bet_id = {0};", bet_id);

            int queryResult;
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();
            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("bet_won", result);
            command.Parameters.AddWithValue("odd", odd);
            command.Parameters.AddWithValue("bet", bet);
            command.Parameters.AddWithValue("name", name);
            try
            {
                queryResult = command.ExecuteNonQuery();
            }
            finally
            {
                con.Close();
            }
            return queryResult;
        }

        /// <summary>
        /// Deletes the bet from table bets that matches bet_id. Returns number of rows affected by the statement.
        /// Deletes also all matches from table bet_in_bet_folder, since bet deletion cascades.
        /// </summary>
        /// <param name="bet_id">id of the bet to be deleted.</param>
        /// <param name="user_id">Id of the user deleting the bet.</param>
        /// <exception cref="UnknownBetError">Thrown when bet with bet_id 'id' is not found.</exception>
        /// <exception cref="AuthenticationError">Thrown when user tries to delete another user's bet.</exception>
        public static int DeleteBet(string ConnectionString, int bet_id, int user_id)
        {
            if (GetBet(ConnectionString, bet_id) == null)
                throw new UnknownBetError("Bet with id " + bet_id + " was not found"); 

            if (GetBet(ConnectionString, bet_id).getOwner() != user_id)
                throw new AuthenticationError("Bet did not belong to user trying to delete it");

            string query = "DELETE FROM bets WHERE bet_id = @bet_id";

            int queryResult;
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();
            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("bet_id", bet_id);
            try
            {
                queryResult = command.ExecuteNonQuery();
            }
            finally
            {
                con.Close();
            }
            return queryResult;
        }

        /// <summary>
        /// Removes a selected bet from selected folders.
        /// </summary>
        /// <exception cref="UnknownBetError">Thrown when bet with bet_id 'bet_id' is not found.</exception>
        /// <exception cref="AuthenticationError">Thrown when user tries to delete another user's bet.</exception>
        public static int DeleteBetFromFolders(string ConnectionString, int bet_id, int user_id, List<string> folders)
        {
            Bet bet = GetBet(ConnectionString, bet_id);
            if (bet == null)
                throw new UnknownBetError("Bet with id " + bet_id + " was not found");

            if (bet.getOwner() != user_id)
                throw new AuthenticationError("Bet did not belong to user trying to delete it");

            if (folders == null || folders.Count == 0)
                return -1;

            string query = "DELETE FROM bet_in_bet_folder WHERE bet_id = @bet_id AND (";

            for (int i = 0; i < folders.Count; i++)
            {
                query = query + String.Format("folder = '{0}' ", folders[i]);

                if (i < folders.Count - 1)
                    query = query + "OR ";
            }
            query = query + ");";

            int queryResult;
            SQLiteConnection con = new SQLiteConnection(ConnectionString);
            con.Open();
            SQLiteCommand command = new SQLiteCommand(query, con);
            command.Parameters.AddWithValue("bet_id", bet_id);
            try
            {
                queryResult = command.ExecuteNonQuery();
            }
            finally
            {
                con.Close();
            }
            return queryResult;
        }
    }
}

File: database/bets.js
var moment = require('moment');
const isNumber = require('is-number');

module.exports = {
  /*
  Returns bets from selected user. Returns only finished betsi
  if {bets_finished} is true, unfinished bets if it is false,
  and all if {bets_finished} is not given at all.

  Returns null on error in request.

  Bet format:
  var bet = {
    bet_won: integer, //-1 = not finished, 0=lost, 1=won
    name: string, //optional name for the bet
    odd: double,
    bet: double,
    date_time: string, //
    owner: integer, //id of the user who owns the bet
    bet_id: integer
  }
  */
  get_bets: function(db_path, user_id, bets_finished){
    const db = require('better-sqlite3')(db_path);
    let result = [];
    try {
      let stmt;
      let res;
      if (bets_finished === undefined){
        stmt = db.prepare('SELECT * FROM bets WHERE owner=?');
        res = stmt.all(user_id);
        result = res.map((item) => { return create_bet_object(item); });
      }
      else {
        let query = bets_finished ? 'SELECT * FROM bets WHERE owner=? AND bet_won != -1' : 'SELECT * FROM bets WHERE owner=? AND bet_won = - 1';
        stmt = db.prepare(query);
        res = stmt.all(user_id);
        result = res.map((item) => { return create_bet_object(item); });
      }
    }
    catch (err){
      result = null;
    }
    finally{
      db.close();
    }
    return result;
  },

  get_bet: function(db_path, bet_id){
    const db = require('better-sqlite3')(db_path);
    let bet;
    try {
      let res = db.prepare('SELECT * FROM bets WHERE bet_id=?').get(bet_id);
      bet = create_bet_object(res);
    }
    catch (err){
      bet = null;
    }
    finally{
      db.close();
    }
    return bet;
  },

  /*
  Returns bets from specific folder by user. If finished argument
  is not given, all bets from folder are returned. If it is true,
  only finished bets from folder are returned: If false, only
  unfinished bets from folder are returned.

  Returns null on error in request.
  */
  get_bets_from_folder: function(db_path, user_id, folder, bets_finished){
    const db = require('better-sqlite3')(db_path);
    let result = [];
    try {
      let stmt;
      let res;
      if (bets_finished === undefined){
        stmt = db.prepare('SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = ? and bf.folder = ?');
      }
      else if(bets_finished) {
        let query = 'SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = ? and bf.folder = ? and bet_won != -1';
        stmt = db.prepare(query);
      }
      else if (!bets_finished){
        let query = 'SELECT * FROM bet_in_bet_folder bf INNER JOIN bets b ON b.bet_id = bf.bet_id WHERE bf.owner = ? and bf.folder = ? and bet_won = -1';
        stmt = db.prepare(query);
      }
      res = stmt.all(user_id, folder);
      result = res.map((item) => { return create_bet_object(item); });
    }
    catch (err){
      result = null;
    }
    finally{
      db.close();
    }
    return result;
  },

  /*
  Delete's a bet from all folders of user.
  Returns true on success, false on failure & null
  on error in request.
  */
  delete_bet: function(db_path, bet_id, user_id){
    let bet_to_delete = this.get_bet(db_path, bet_id);
    if (bet_to_delete === null || bet_to_delete.owner !== user_id){
      return false;
    }

    const db = require('better-sqlite3')(db_path);
    let query = 'DELETE FROM bets WHERE bet_id = ?';
    let result = false;
    try {
      var res = db.prepare(query).run(bet_id);
      if (res.changes === 1){
        result = true;
      }
    }
    catch (err){
      result = null;
    }
    finally{
      db.close();
    }
    return result;
  },

  /*
  Deletes a bet from selected folders. Returns an array
  with names of folders from which bet was successfully deleted.

  Returns null when a connection related error happens.
  */
  delete_bet_from_folders: function(db_path, folders, bet_id, user_id){
    if (this.get_bet(db_path, bet_id).owner !== user_id){
      return [];
    }

    const db = require('better-sqlite3')(db_path);
    let query = 'DELETE FROM bet_in_bet_folder WHERE bet_id = ? AND folder = ? ';
    let deletedFrom = [];
    let stmt = db.prepare(query);
    for (var i = 0; i < folders.length; i++){
      try {
        let res = stmt.run(bet_id, folders[i]);
        if (res.changes === 1){
          deletedFrom.push(folders[i]);
        }
      }
      catch(err){
        if (err.message.indexOf('no such table') !== -1){
          deletedFrom = null;
          break;
        }
      }
    }
    db.close();

    return deletedFrom;
  },

  /*
  Creates a new bet.

  parameters:
    db_path: Path to database file_operations.
    user_id: Id for the user who played the bet.
    datetime: Datetime when the bet was played.
    odd: Odd for the bet.
    bet: Stake for the bet.
    name: Name for the bet.
    bet_won: true if bet was correct, false if not,
      null if bet has not resolved yet.

  Returns true on success,
  false if user does not exist, or if datetime is in
  invalid format, and null if request fails because
  of database and connection issues.
  */
  create_bet: function(db_path, user_id, datetime, odd, bet, name, bet_won){
    datetime = moment(datetime, 'YYYY-MM-DD HH:MM:SS').format('YYYY-MM-DD HH:MM:SS');
    if(!moment(datetime, 'YYYY-MM-DD HH:MM:SS', true).isValid()){
      return false;
    }

    if(!isNumber(odd) || !isNumber(bet)){
      return false;
    }

    let bet_result = -1;
    if (bet_won !== null){
      bet_result = bet_won ? 1 : 0;
    }

    const db = require('better-sqlite3')(db_path);
    let query = 'INSERT INTO bets (bet_won, name, odd, bet, date_time, owner) values (?, ?, ?, ?, ?, ?)';
    let result = false;
    try {
      var res = db.prepare(query).run(bet_result, name, odd, bet, datetime, user_id);

      if (res.changes === 1){
        result = true;
      }
    }
    catch (err){
      if (err.message.indexOf('no such table') === -1){
        result = false;
      }
      else {
        result = null;
      }
    }
    finally{
      db.close();
    }
    return result;
  },

  /*
  Add's an already existing bet into folders.
  Returns an array of folder names, to which bet
  was successfully added.

  Returns null when an error regarding connection is found.
  */
  add_bet_to_folders: function(db_path, folders, bet_id, user_id){
    if (this.get_bet(db_path, bet_id).owner !== user_id){
      return [];
    }

    const db = require('better-sqlite3')(db_path);
    let query = 'INSERT INTO bet_in_bet_folder VALUES (?, ?, ?)';
    let addedTo = [];
    let stmt = db.prepare(query);
    for (var i = 0; i < folders.length; i++){
      try {
        let res = stmt.run(folders[i], user_id, bet_id);
        if (res.changes === 1){
          addedTo.push(folders[i]);
        }
      }
      catch(err){
        if (err.message.indexOf('no such table') !== -1){
          addedTo = null;
          break;
        }
      }
    }
    db.close();

    return addedTo;
  }
}

function create_bet_object(item){
  var bet = {
    bet_won: item.bet_won,
    name: item.name,
    odd: item.odd,
    bet: item.bet,
    date_time: item.date_time,
    owner: item.owner,
    bet_id: item.bet_id
  }
  return bet;
}
